src/noflo-component-access.js

```diff
@@ -34,75 +36,14 @@ function facadePort(nodeInstance, port, name) {
     return _.extend({
         name: name,
         nodeInstance: nodeInstance,
-        isMulti: port.isAddressable.bind(port)
-    }, facadeFunctions(port,
-        'isAddressable',
-        'isRequired',
-        'listAttached',
-        'connect',
-        'send',
-        'disconnect'
-    ), facadeEventEmitter(port));
-}
-
-/**
- * Returns a facade of the given EventEmitter. The result is not an EventEmitter,
- * but includes the same functions. Each function call is passed to the underlying
- * EventEmitter, while any provided functions are bound to the original context
- * when they are called. This gives the behaviour of an EventEmitter without
- * actually creating one with a distinct state.
- * @see https://nodejs.org/dist/latest-v4.x/docs/api/events.html
- */
-function facadeEventEmitter(emitter) {
-    return facadeFunctions(emitter,
-        'emit',
-        'on',
-        'once',
-        'removeListener'
-    );
-}
-
-/**
- * Picks all the given functions from object and returns the set bounded to the
- * original object context. If any argument is passed to those functions that is
- * also a function it is bound to the called context.
- * @param object with functions, i.e. noflo.Component, noflo.InPort/OutPort
- * @param keys explicit names of functions to include in result
- */
-function facadeFunctions(object /* keys */) {
-    var keys = _.toArray(arguments).slice(1);
-    return _.mapObject(_.pick(object, keys), function(fn) {
-        return _.partial(applyFacade, fn, object)
-    });
-}
-
-/**
- * If any of the arguments are functions, replace them with a bound version
- * before calling fn.
- * @this context that optional function arguments should be bound to
- * @param fn a function that might take other functions as arguments
- * @param context that fn should be called with
- */
-function applyFacade(fn, context /* arguments */) {
-    var args = _.toArray(arguments).slice(2);
-    return fn.apply(context, _.map(args, facadeFunction, this));
-}
-
-/**
- * Binds the given function to this context or returns the arguments if not a
- * function. This function will return the same result for repeated calls.
- * @this context to bind the function to
- * @param fn a function to bind
- */
-function facadeFunction(fn) {
-    if (!_.isFunction(fn)) return fn;
-    if (!_.isArray(fn._bind_facade_noflo_access)) {
-        fn._bind_facade_noflo_access = [];
-    }
-    var idx = fn._bind_facade_noflo_access.indexOf(this);
-    if (idx >= 0) return fn._bind_facade_noflo_access[idx + 1];
-    fn._bind_facade_noflo_access.push(this, fn.bind(this));
-    return _.last(fn._bind_facade_noflo_access);
+        isMulti: port.isAddressable.bind(port),
+        isRequired: port.isRequired.bind(port),
+        listAttached: port.listAttached.bind(port)
+    }, isOutPort(port) ? {
+        connect: port.connect.bind(port),
+        send: port.send.bind(port),
+        disconnect: port.disconnect.bind(port)
```

Comment:
I would rather not expose connect, send and disconnect to the RDF Pipeline layer.  Can you expose only a sendState function instead, which does the noflo connect-send-disconnect sequence?  Or do you temporarily need access to the individual connect, send and disconnect functions for some of the existing test code?